### PR TITLE
Name allowed keys and/guess on missing map key

### PR DESCRIPTION
**Pull request by pivotaljohn**
Replaces existing author-centered hint with some that help the consumer, instead.

Inspired by:

image

cc: @voor

### CODE REVIEW

File: pkg/schema/error.go [hunk 1]
```diff
@@ -95,13 +97,28 @@ func NewMismatchedTypeAssertionError(foundType yamlmeta.TypeWithValues, expected
...
+// NewUnexpectedKeyAssertionError generates a schema assertion error including the context (and hints) needed to report it to the user
+func NewUnexpectedKeyAssertionError(found *yamlmeta.MapItem, definition *filepos.Position, allowedKeys []string) error {
+	key := fmt.Sprintf("%v", found.Key)
+	err := schemaAssertionError{
+		description: "Given data value is not declared in schema",
+		position:    found.GetPosition(),
+		found:       key,
+	}
+	sort.Strings(allowedKeys)
+	switch numKeys := len(allowedKeys); {
```
Comment (cppforlife) [inline pkg/schema/error.go:106]: style nit: no need for `;`

File: pkg/schema/error.go [hunk 2]
```diff
@@ -95,13 +97,28 @@ func NewMismatchedTypeAssertionError(foundType yamlmeta.TypeWithValues, expected
...
+	case numKeys == 1:
+		err.expected = fmt.Sprintf("a %s with the key named \"%s\" (from %s)", found.DisplayName(), allowedKeys[0], definition.AsCompactString())
+	case numKeys > 1 && numKeys <= 9: // Miller's Law
```
Comment (cppforlife) [inline pkg/schema/error.go:109]: love this!

File: pkg/schema/error.go [hunk 3]
```diff
@@ -95,13 +97,28 @@ func NewMismatchedTypeAssertionError(foundType yamlmeta.TypeWithValues, expected
...
+		err.expected = fmt.Sprintf("one of { %s } (from %s)", strings.Join(allowedKeys, ", "), definition.AsCompactString())
+	default:
+		err.expected = fmt.Sprintf("(a key declared in map, in schema at %s)", definition.AsCompactString())
+	}
+	mostSimilarKey := spell.Nearest(key, allowedKeys)
+	if mostSimilarKey != "" {
+		err.hints = append(err.hints, fmt.Sprintf("did you mean \"%s\"?", mostSimilarKey))
```
Comment (cppforlife) [inline pkg/schema/error.go]: style nit: can use tick

File: pkg/spell/spell.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package spell file defines a simple spelling checker for use in attribute errors
+// such as "no such field .foo; did you mean .food?".
```
Comment (cppforlife) [inline pkg/spell/spell.go:5]: not sure if we took this file from starlark. if we did, lets mention that in a comment

File: pkg/schema/error.go (hunk 1 above)
Comment (pivotaljohn) [reply pkg/schema/error.go:106]: This is required given that an assignment is not an expression.

The exact error message:
> pkg/schema/error.go:107:9: expected switch expression, found assignment (missing parentheses around composite literal?)

File: pkg/schema/error.go (hunk 2 above)
Comment (pivotaljohn) [reply pkg/schema/error.go:109]: 🙇🏻

File: pkg/schema/error.go (hunk 3 above)
Comment (pivotaljohn) [reply pkg/schema/error.go]: 🙏🏻

File: pkg/spell/spell.go (hunk 4 above)
Comment (pivotaljohn) [reply pkg/spell/spell.go:5]: We did. I will.

**Comment by pivotaljohn** [comment]
Tests failing in CI, but working locally: I failed to rebase off `develop`; on it.

**Review by cari-lynn — APPROVED** [review]
🥇

**Pull request merged**